src/Servers/Kestrel/Transport.MsQuic/Internal/MsQuicStream.cs

```diff
@@ -0,0 +1,241 @@
+                    status = HandleStartComplete(stream);
+                    break;
+                case QUIC_STREAM_EVENT.RECV:
+                    {
+                        status = HandleEventRecv(
+                            stream,
+                            ref evt);
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                    {
+                        status = HandleEventSendComplete(ref evt);
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.PEER_SEND_CLOSE:
+                    {
+                        status = HandleEventPeerSendClose();
+                    }
+                    break;
+                // TODO figure out difference between SEND_ABORT and RECEIVE_ABORT
+                case QUIC_STREAM_EVENT.PEER_SEND_ABORT:
+                    {
+                        _streamClosedTokenSource.Cancel();
+                        status = HandleEventPeerSendAbort();
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.PEER_RECV_ABORT:
+                    {
+                        _streamClosedTokenSource.Cancel();
+                        status = HandleEventPeerRecvAbort();
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.SEND_SHUTDOWN_COMPLETE:
+                    {
+                        status = HandleEventSendShutdownComplete(ref evt);
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.SHUTDOWN_COMPLETE:
+                    {
+                        _stream.Close();
+                        return QUIC_STATUS.SUCCESS;
+                    }
+
+                default:
+                    break;
+            }
+            return status;
+        }
+
+        private QUIC_STATUS HandleEventPeerRecvAbort()
+        {
+            return QUIC_STATUS.SUCCESS;
+        }
+
+        private QUIC_STATUS HandleEventPeerSendAbort()
+        {
+            return QUIC_STATUS.SUCCESS;
+        }
+
+        private QUIC_STATUS HandleStartComplete(QuicStream stream)
+        {
+            return QUIC_STATUS.SUCCESS;
+        }
+
+        private QUIC_STATUS HandleEventSendShutdownComplete(ref NativeMethods.StreamEvent evt)
+        {
+            return QUIC_STATUS.SUCCESS;
+        }
+
+        private QUIC_STATUS HandleEventPeerSendClose()
+        {
+            // TODO complete async
+            // Close as fast as possible here.
+            Input.Complete();
+            return QUIC_STATUS.SUCCESS;
+        }
+
+        public QUIC_STATUS HandleEventSendComplete(ref NativeMethods.StreamEvent evt)
+        {
+            return QUIC_STATUS.SUCCESS;
+        }
+
+        protected QUIC_STATUS HandleEventRecv(QuicStream stream, ref NativeMethods.StreamEvent evt)
+        {
+            var input = Input;
+            var length = (int)evt.TotalBufferLength;
```

Comment:
@jkotalik If possible, I would do a Write here manually instead of asking for a single buffer of `TotalBufferLength`. That'll ensure it's using the memory pool instead of potentially allocating